### PR TITLE
Fix bug + Add Tests + Enhance docstrings (shape_equal)

### PR DESCRIPTION
**Pull request by Faisal-Alsrheed**
**Bug Fix:**

In `shape_equal `function, using `axis=1` resulted in the `error: TypeError: 'int' object is not iterable.`
but  `axis=[1] `worked correctly.

After hours of trying  :)  the issue has been resolved, and now an integer, list, or tuple of integers  work as expected and documented.

**Improved Docstrings:**
The documentation for the shape_equal function has been enhanced. It now provides a more in-depth description and includes illustrative examples.

**Additional Tests:**

A new set of tests has been added to ensure the proper functionality of the `shape_equal `function.
Why do we need tests for the `shape_equal` function? Because there are 7 other functions depend on it:

1. Append ( )
2. Average ( )
3. Concatenate ( )
4. Hstack ( )
5. Stack ( )
6. Tensordot ( )
7. Vstack () 

Please share your feedback, thank you for the guidance, I am learning so much :)

### CODE REVIEW

File: keras_core/ops/numpy.py [hunk 1]
```diff
@@ -197,24 +197,53 @@ def broadcast_shapes(shape1, shape2):
 
 
 def shape_equal(shape1, shape2, axis=None, allow_none=True):
-    """Check if two shapes are equal.
+    """Check if two shapes are equal
```
Comment (fchollet) [inline keras_core/ops/numpy.py:200]: Add back the period.

File: keras_core/ops/numpy.py [hunk 2]
```diff
@@ -197,24 +197,53 @@ def broadcast_shapes(shape1, shape2):
...
-        shape2: A tuple or list of integers.
-        axis: int or list/tuple of ints, defaults to `None`. If specified, the
-            shape check will ignore the axes specified by `axis`.
-        allow_none: bool, defaults to `True`. If `True`, `None` in the shape
-            will match any value.
+        shape1: A list or tuple of integers for first shape to be compared.
+        shape2: A list or tuple of integers for second shape to be compared.
+        axis: An integer, list, or tuple of integers (optional):
+            Axes to ignore during comparison.Default is None.
+        allow_none (bool, optional): If True, allows None in a shape to match
```
Comment (fchollet) [inline keras_core/ops/numpy.py:207]: Use backticks for True & None

File: keras_core/ops/numpy.py [hunk 3]
```diff
@@ -197,24 +197,53 @@ def broadcast_shapes(shape1, shape2):
...
-        shape1: A tuple or list of integers.
-        shape2: A tuple or list of integers.
-        axis: int or list/tuple of ints, defaults to `None`. If specified, the
-            shape check will ignore the axes specified by `axis`.
-        allow_none: bool, defaults to `True`. If `True`, `None` in the shape
-            will match any value.
+        shape1: A list or tuple of integers for first shape to be compared.
+        shape2: A list or tuple of integers for second shape to be compared.
+        axis: An integer, list, or tuple of integers (optional):
+            Axes to ignore during comparison.Default is None.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:206]: Add space after period

File: keras_core/ops/numpy.py [hunk 4]
```diff
@@ -197,24 +197,53 @@ def broadcast_shapes(shape1, shape2):
...
+            any value in the corresponding position of the other shape.
+            Default is True.
+
+    Returns:
+        bool: True if shapes are considered equal based on the criteria,
+        False otherwise.
+
+    Examples:
+
+        >>> shape_equal((32, 64, 128), (32, 64, 128))
```
Comment (fchollet) [inline keras_core/ops/numpy.py:217]: Unindent code example

**Review by fchollet — APPROVED** [review]
Thanks for the PR!

**Comment by Faisal-Alsrheed** [comment]
Hi @fchollet :)
 
Which Python function docstring standard does the Keras team use?

Please recommend a python library for docstring. Black is not working with docstring as far as u know.

I can not find any information
https://www.perplexity.ai/search/054000c8-6e38-427d-bed7-930814e4f4ed?s=u

This ?
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings

This ?
https://github.com/keras-team/keras-cv/blob/5373b916d15544a6763347575d440570ec617495/.github/CONTRIBUTING.md#formatting-the-code

Thank you

File: keras_core/ops/numpy.py [hunk 5]
```diff
@@ -200,21 +200,50 @@ def shape_equal(shape1, shape2, axis=None, allow_none=True):
...
-        shape1: A tuple or list of integers.
-        shape2: A tuple or list of integers.
-        axis: int or list/tuple of ints, defaults to `None`. If specified, the
-            shape check will ignore the axes specified by `axis`.
-        allow_none: bool, defaults to `True`. If `True`, `None` in the shape
-            will match any value.
+        shape1: A list or tuple of integers for first shape to be compared.
+        shape2: A list or tuple of integers for second shape to be compared.
+        axis: An integer, list, or tuple of integers (optional):
+            Axes to ignore during comparison. Default is 'None'.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:206]: Use backticks rather than quotes around code keywords, e.g. `

**Comment by fchollet** [comment]
> Black is not working with docstring as far as u know.

You can copy/paste code snippets into a black formatting tool, there are webpages that can do it.

**Review by fchollet — APPROVED** [review]
LGTM, thank you!

**Pull request merged**